File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = {}
        l = 0
        res = 0

        for r in range(len(s)):
            if s[r] in mp:
                l = max(mp[s[r]]+1,l) #保證左指標前進，取max原因是假設abba，直接取mp[s[r]]+1左指標會往前跑
            mp[s[r]] = r
            res = max(res,r-l+1)
        return res
        

File: Data Structures & Algorithms/sqrtx/submission-0.java
class Solution {
    public int mySqrt(int x) {
        int res = 0;
        int l = 1;
        int r = x;

        if (x == 0)
        {
            return 0;
        }

        while (l <= r)
        {
            int m = l + (r - l)/2;
            if ((long) m * m > x)
            {
                r = m - 1;
            }
            else if ((long)m * m < x)
            {
                l = m + 1;
                res = m;
            }
            else
            {
                return m;
            }
        }
        return res;
    }
}